addon/lib/field.js

```diff
@@ -207,6 +208,130 @@ export default Base.extend({
    */
   value: reads("answer.value"),
 
+  /**
+   * Fetch all formerly used dynamic options for this question. This will be
+   * taken from the apollo cache if possible.
+   *
+   * @method _fetchUsedDynamicOptions.perform
+   * @return {Object[]} Formerly used dynamic options
+   * @private
+   */
+  _fetchUsedDynamicOptions: task(function*() {
+    if (!this.question.isDynamic) return null;
+
+    const edges = yield this.apollo.query(
+      {
+        query: getDocumentUsedDynamicOptionsQuery,
+        fetchPolicy: "cache-first",
+        variables: {
+          document: this.document.uuid,
+          question: this.question.slug
+        }
+      },
+      "allUsedDynamicOptions.edges"
+    );
+
+    return edges.map(({ node: { slug, label } }) => ({
+      slug,
+      label
+    }));
+  }),
+
+  /**
+   * The formerly used dynamic options for this question.
+   *
+   * @property {Object[]} usedDynamicOptions
+   * @accessor
+   *
+   */
+  usedDynamicOptions: reads("_fetchUsedDynamicOptions.lastSuccessful.value"),
+
+  /**
+   * The available options for choice questions. This only works for the
+   * following types:
+   *
+   * - ChoiceQuestion
+   * - DynamicChoiceQuestion
+   * - MultipleChoiceQuestion
+   * - DynamicMultipleChoiceQuestion
+   *
+   * This will also return the disabled state of the option. An option can only
+   * be disabled, if it is an old value used in a dynamic question.
+   *
+   * @property {Null|Object[]} options
+   * @accessor
+   */
+  options: computed(
+    "value",
+    "question.options.[]",
+    "usedDynamicOptions.[]",
+    function() {
+      if (!this.question.isChoice && !this.question.isMultipleChoice) {
+        return null;
+      }
+
+      const selected =
+        (this.question.isMultipleChoice ? this.value : [this.value]) || [];
+
+      const options = this.question.options.map(option => ({
+        ...option,
+        disabled: false
+      }));
+
+      const hasInvalidSelected = !selected.every(slug =>
+        options.find(option => option.slug === slug)
+      );
+
+      if (this.question.isDynamic && hasInvalidSelected) {
+        if (!this._fetchUsedDynamicOptions.lastSuccessful) {
+          // Fetch used dynamic options if not done yet
+          this._fetchUsedDynamicOptions.perform();
```

Comment:
Shouldn't we wait here? (await or yield)

Reply:
This is a computed property, we can't await here

Reply:
But don't we risk that `this.usedDynamicOptions` is still fetching when we enter the return statement a few lines below?

Reply:
What would be the risk of that?

Reply:
Probably, unexpected behavior. But I don't know. That's why I ask.

Reply:
It's actually intended behaviour which will return the options without any used options. Since it will be recomputed "correctly" shortly after, this should not cause any problems.

Reply:
Ok